### PR TITLE
array-to-raw-elem cast: test that Retag covers entire array

### PR DESCRIPTION
**Pull request by RalfJung**
Make sure that we `Retag` *before* doing the `ArrayToPointer` cast.

### CODE REVIEW

File: src/test/mir-opt/retag.array_casts.SimplifyCfg-elaborate-drops.after.mir [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+        StorageLive(_1);                 // scope 0 at $DIR/retag.rs:58:9: 58:14
+        _1 = [const 0_usize, const 0_usize]; // scope 0 at $DIR/retag.rs:58:29: 58:35
+        StorageLive(_2);                 // scope 1 at $DIR/retag.rs:59:9: 59:10
+        StorageLive(_3);                 // scope 1 at $DIR/retag.rs:59:13: 59:19
+        StorageLive(_4);                 // scope 1 at $DIR/retag.rs:59:13: 59:19
+        _4 = &mut _1;                    // scope 1 at $DIR/retag.rs:59:13: 59:19
+        Retag(_4);                       // scope 1 at $DIR/retag.rs:59:13: 59:19
+        _3 = &raw mut (*_4);             // scope 1 at $DIR/retag.rs:59:13: 59:19
+        Retag([raw] _3);                 // scope 1 at $DIR/retag.rs:59:13: 59:19
+        _2 = move _3 as *mut usize (Pointer(ArrayToPointer)); // scope 1 at $DIR/retag.rs:59:13: 59:33
```
Comment (RalfJung) [inline src/test/mir-opt/retag.array_casts.SimplifyCfg-elaborate-drops.after.mir:74]: This is the key part of the test. I am not sure how to make sure this won't be bless'ed away...

Comment (Mark-Simulacrum) [reply src/test/mir-opt/retag.array_casts.SimplifyCfg-elaborate-drops.after.mir:74]: Hm. I wonder if we want something like LLVM's FileCheck for mir-opt tests... cc @rust-lang/wg-mir-opt do we have any standard practice here?

**Comment by Mark-Simulacrum** [comment]
r=me in theory, though I agree this seems error-prone as-is.

**Comment by RalfJung** [comment]
@oli-obk I recall you working on this infrastructure so that mir-opt tests can be blessed; is there any way to make sure that the test actually keeps testing what it is supposed to test though? The retag mir-opt test as it exists currently in master is a lot more fragile that I'd like it to be (and back when I added the test, it was less fragile, back then it *did* ensure that there'd be `Retag`s at the right places -- something got removed in the mean time).

**Comment by oli-obk** [comment]
There is no existing infrastructure, but it would be possible to add. I think the hard part is knowing what we actually want.

We could do something simple like requiring that a certain string exists in the mir dump, that could easily be added to compiletest by adding more magic comments like the ones for actually doing a dump.

**Pull request merged**